crates/tendermint/src/block/id.rs

```diff
@@ -0,0 +1,122 @@
+use digest::{
+    generic_array::{typenum::Unsigned, GenericArray},
+    Digest,
+};
+use ripemd160::Ripemd160;
+#[cfg(feature = "serializers")]
+use serde::{de::Error as DeError, Deserialize, Deserializer, Serialize, Serializer};
+use std::{
+    fmt::{self, Debug, Display},
+    str::{self, FromStr},
+};
+use subtle_encoding::hex;
+
+use error::Error;
+
+/// Block identifiers
+#[derive(Clone, Hash, Eq, PartialEq)]
+pub struct Id<D>
+where
+    D: Digest,
+{
+    /// Hash which identifies this block
+    pub hash: GenericArray<u8, D::OutputSize>, // TODO: parts set header?
+}
+
+impl<D> Id<D>
+where
+    D: Digest,
+{
+    /// Create a new `Id` from a hash byte slice
+    pub fn new(hash: &[u8]) -> Result<Self, Error> {
+        if hash.len() != D::OutputSize::to_usize() {
+            return Err(Error::Parse);
+        }
+
+        Ok(Self {
+            hash: GenericArray::clone_from_slice(hash),
+        })
+    }
+}
+
+impl<D> Debug for Id<D>
+where
+    D: Digest,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "block::Id({})", self.to_string())
+    }
+}
+
+// TODO: match gaia serialization? e.g `D2F5991B98D708FD2C25AA2BEBED9358F24177DE:1:C37A55FB95E9`
+impl<D> Display for Id<D>
+where
+    D: Digest,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            String::from_utf8(hex::encode_upper(self.hash.as_slice())).unwrap()
+        )
+    }
+}
+
+// TODO: match gaia serialization?
+impl<D> FromStr for Id<D>
+where
+    D: Digest,
+{
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Error> {
+        Self::new(&hex::decode_upper(s)?)
+    }
+}
+
+#[cfg(feature = "serializers")]
+impl<D> Serialize for Id<D>
+where
+    D: Digest,
+{
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        self.to_string().serialize(serializer)
+    }
+}
+
+#[cfg(feature = "serializers")]
+impl<'de, Dg> Deserialize<'de> for Id<Dg>
+where
+    Dg: Digest,
+{
+    fn deserialize<De: Deserializer<'de>>(deserializer: De) -> Result<Self, De::Error> {
+        Self::from_str(&String::deserialize(deserializer)?)
+            .map_err(|e| De::Error::custom(format!("{}", e)))
+    }
+}
+
+/// Block IDs computed using the RIPEMD-160 hash function
+pub type Ripemd160Id = Id<Ripemd160>;
```

Comment:
I kind of just assumed the IDs were RIPEMD-160 for now... I believe @ebuchman said something about migrating to SHA-256 everywhere (which I'd recommend). Anyway this code is written in such a way that the IDs are generic over the underlying hash function, so switching to should be easy.

Reply:
We got rid of RIPEMD some time ago - it's SHA256[:20] everywhere now. Though soon it will be full SHA256 everywhere, except for ED25519 addresses, which will remain SHA256[:20]. See https://github.com/tendermint/tendermint/issues/1990#issuecomment-431392096 for summary